src/groups/mqb/mqbblp/mqbblp_cluster.h

```diff
@@ -311,6 +311,8 @@ class Cluster : public mqbi::Cluster,
     // responses from proxies and nodes,
     // and the cluster's shutdown callback.
 
+    mqbnet::Session::AdminCommandEnqueueCb d_adminCb;
```

Comment:
I wonder if this is necessary.  Cluster receives an admin command (is this always the rerouted one?).  Do we need to enqueue the command to the Application, which then looks up cluster and ends up calling the (same) cluster?

Do we want to support cluster receiving commands for another cluster?

This may worth a discussion.  Let's zoom one day?

Reply:
The idea is that all commands get processed through the application because that is where we have the necessary dependencies to execute a command (`d_domainManager_mp`, `d_configProvider_mp`, etc.).

A command that is rerouted should always invoke that callback, `d_adminCb`, when the command is received, and that callback enqueues the command to the application instance where it will be executed. A command that has been rerouted never performs additional routing, it just gets executed by the application.

> Do we want to support cluster receiving commands for another cluster?

I'm not sure what exactly you mean by this.

I think a zoom is a good idea

Reply:
> > Do we want to support cluster receiving commands for another cluster?
> 
> I'm not sure what exactly you mean by this.

I meant, can cluster "X" receives a (cluster) command (in `Cluster::processControlMessage`) that ends up being processed by cluster "Y"?

What commands are "re-routable"?  From `CommandRouter::getCommandRoutingMode()`: `isDomainsValue()` and `isClustersValue()`.

I am just pointing out that an object of our our routing is a cluster (`getRelevantCluster`) and that is the cluster receiving  the rerouted command (in `Cluster::processControlMessage`).  I wonder if `Application` code can be simplified by not dealing with rerouted commands if the cluster executes rerouted commands when it receives them.   Provided, clusters can parse the commands.

Reply:
Oh I see! No, we do not support a cluster executing a command for another cluster. So we may be able to move some of that logic to the cluster code.

The problem, however, is that a lot of the command processing code lives in the `Application` currently (parsing a command, executing it, and formatting results). Some of this could be abstracted to another file/class and be reused by the application and cluster code, but the execution logic needs to live in the `Application` since it utilizes many high-level dependencies. Doing this might make the Application code simpler but would introduce some complexity elsewhere.

Reply:
I think, there is a possibility of routing commands which are _not_ cluster related, in which case it the `Application` must process the command.  So, calling `Application` callback is the preferred way after all.